# TUGAS PRAKTIKUM 2

**Load Data**

In [ ]:
import pandas as pd
data = pd.read_csv('/content/drive/MyDrive/Machine Learning   15/Data/spam.csv', encoding='latin-1')
data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

**Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,**

**Menggunakan data spam.csv**

**Fitur CountVectorizer dengan mengaktifkan stop_words**

**Evaluasi hasilnya**

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Load data
data = pd.read_csv('/content/drive/MyDrive/Machine Learning   15/Data/spam.csv', encoding='latin-1')

# Preprocessing: Drop unnecessary columns and encode labels
# Drop 3 kolom terakhir dengan fungsi iloc
data = data.drop(data.iloc[:,2:], axis=1)

# Data untuk label
new_labels = {
    'spam': 1,
    'ham': 0
}

# Encode label
data['v1'] = data['v1'].map(new_labels)

# Pisahkan fitur dengan label
X = data['v2'].values
y = data['v1'].values


# Split data training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

# Inisiasi CountVectorizer dengan stop_words
bow = CountVectorizer(stop_words='english')

# Fitting dan transform X_train dengan CountVectorizer
X_train = bow.fit_transform(X_train)

# Transform X_test
X_test = bow.transform(X_test)

# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data testing
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data testing
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_test))
print("\nLaporan Klasifikasi:\n", classification_report(y_test, y_pred_test))

Hasil akurasi data train: 0.9946152120260264
Hasil akurasi data test: 0.9829596412556054

Confusion Matrix:
 [[951   3]
 [ 16 145]]

Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.98      1.00      0.99       954
           1       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



**Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,**

**Menggunakan data spam.csv**

**Fitur TF-IDF dengan mengaktifkan stop_words**

**Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.**

**Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv**

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Pisahkan fitur dengan label
X = data['v2'].values
y = data['v1'].values

# Split data training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

# Inisiasi TfidfVectorizer dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')

# Fitting dan transform X_train dengan TfidfVectorizer
X_train_tfidf = tfidf.fit_transform(X_train)

# Transform X_test
X_test_tfidf = tfidf.transform(X_test)

# Inisiasi MultinomialNB
mnb_tfidf = MultinomialNB()

# Fit model
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi dengan data training
y_pred_train_tfidf = mnb_tfidf.predict(X_train_tfidf)

# Evaluasi akurasi data training
acc_train_tfidf = accuracy_score(y_train, y_pred_train_tfidf)

# Prediksi dengan data testing
y_pred_test_tfidf = mnb_tfidf.predict(X_test_tfidf)

# Evaluasi akurasi data testing
acc_test_tfidf = accuracy_score(y_test, y_pred_test_tfidf)

# Print hasil evaluasi
print(f'Hasil akurasi data train (TF-IDF): {acc_train_tfidf}')
print(f'Hasil akurasi data test (TF-IDF): {acc_test_tfidf}')
print("\nConfusion Matrix (TF-IDF):\n", confusion_matrix(y_test, y_pred_test_tfidf))
print("\nLaporan Klasifikasi (TF-IDF):\n", classification_report(y_test, y_pred_test_tfidf))

Hasil akurasi data train (TF-IDF): 0.9842943684092439
Hasil akurasi data test (TF-IDF): 0.9605381165919282

Confusion Matrix (TF-IDF):
 [[954   0]
 [ 44 117]]

Laporan Klasifikasi (TF-IDF):
               precision    recall  f1-score   support

           0       0.96      1.00      0.98       954
           1       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



**Kesimpulan Tugas 2:**

Berdasarkan percobaan klasifikasi spam menggunakan model Multinomial Naive Bayes dengan dua metode ekstraksi fitur (CountVectorizer dan TF-IDF) yang mengaktifkan stop_words, didapatkan hasil sebagai berikut:

- **CountVectorizer:**
  - Akurasi data train: 0.9946
  - Akurasi data test: 0.9830

- **TF-IDF:**
  - Akurasi data train: 0.9843
  - Akurasi data test: 0.9605

Dari hasil tersebut, dapat disimpulkan bahwa **CountVectorizer memberikan performa yang sedikit lebih baik** dibandingkan TF-IDF pada dataset spam.csv ini, dilihat dari nilai akurasi pada data testing yang lebih tinggi. Ini menunjukkan bahwa frekuensi kemunculan kata (seperti yang ditangkap oleh CountVectorizer) mungkin lebih informatif untuk membedakan spam dan ham pada dataset ini, meskipun TF-IDF juga memperhitungkan pentingnya kata dalam seluruh korpus.